examples/linux64/hello-make/hello.c

```diff
@@ -0,0 +1,51 @@
+// Currently, it's not possible to specify executable entrypoint to ppci
+// linker - so far, the entrypoint is at the beginning of the code segment.
+// And functions appear in the code segment in the order of appearance in
+// the source file, even if "apperance" is merely a prototype. Thus, the
+// first function should be the "startup" function, and it should use
+// inline asm to call other functions (because currently with PPCI, it's
+// not possible to call a function without prototype, and putting a
+// prototype before the startup function will push it down the code segment,
+// precluding from running as entrypoint).
+int _startup()
+{
+    asm("call main");
+    asm("jmp exit");
+}
+
+
+void puts(char *s);
+void putc(char c);
+void syscall(int nr, int a, int b, int c);
+
+int main()
+{
+    puts("Hello, World!");
+}
+
+void puts(char *s)
+{
+    while (*s) {
+        putc(*s++);
+    }
+    putc('\n');
+}
+
+void putc(char c)
+{
+    syscall(1, 1, (int)&c, 1);
+}
+
+void exit()
+{
+    syscall(60, 0, 0, 0);
+}
+
+void syscall(int nr, int a, int b, int c)
+{
+    asm("mov rax, rdi"); // abi param 1
```

Comment:
I added input parameters and clobber info to assembly code, so this should work as well:

```c
asm(
  "mov rax, %0 \n"
  "mov rdi, %1 \n"
  "mov rsi, %2 \n"
  "mov rdx, %3 \n"
  "syscall \n"
:
: "r" (nr), "r" (a), "r" (b), "r" (c) // input registers, patched into %0 .. %3
: "rax", "rdi", "rsi", "rdx"  // clobber registers, handled by the register allocator
);
```

In this example, it might not make much sense, since we target linux anyways, but the code now makes some assumptions about the calling convention of the syscall function itself.

Off course, you might want to cross check the result of this with `tools/ppci_explorer.py`

Reply:
> In this example, it might not make much sense, since we target linux anyways, but the code now makes some assumptions about the calling convention of the syscall function itself.

Yeah, I guess where it comes into play is with optimizations and surrounding code. But it's definitely best practice to use it, and we want to show just that as example, so applying your suggested code verbatim.